exercises/practice/pig-latin/.meta/lib/example.dart

```diff
@@ -0,0 +1,34 @@
+class PigLatin {
```

Comment:
Thank you for the detailed feedback! Agreed that it doesn't make sense to encapsulate a single pure function in a class, I was just following along with what I had seen in the other exercises - I can go ahead and remove the class and make it a top-level function with a qualified import in the test file.

With this change, should `lib/pig_latin.dart` be an empty file with a comment, or should there be a stub of the translate function?

For example:

```dart
String translate(String phrase) {
  // Your code here.
}
```

Reply:
Let it be empty. If you add a stub then the import with the prefix is a hidden feature that the end-learner may not get exposed. Whereas if you leave the file empty, they might have to fiddle around a little to figure out how dart import [prefixes](https://dart.dev/guides/language/language-tour#libraries-and-visibility) work. Which is good.

Reply:
One of the painful bits of learning we've taken over the years is that it's generally counter-productive to make people flail around to discover _something specific_.

What can be productive is when there are a lot of different ways you could approach something, and they can play around with digging deeper into one or more of them.

When there's only one way, and you _have_ to figure out the exact thing, but nothing is guiding you, it often feels like you're being tricked into feeling stupid. If you really would like to use the prefixed import, then I recommend adding some sort of pointer in the right direction so that people at least know what they're looking for. (Maybe `.docs/introduction.append.md` or `.docs/hints.md`).

In this case, if you think that the prefixed imports is important for people to know specifically, it would be a good case for creating a concept exercise that teaches it specifically.

Reply:
> While it is consistent with other exercises, a class encapsulating only 1 pure public function does not provide any benefit. Especially when it does not have any internal state to keep track of.

One thing that occurred to me as I was falling asleep last night, is that this feedback is specific to the example solution.

Our rule of thumb is that the ideal practice exercise has multiple idiomatic approaches, and the test suite should not dictate a narrow approach, but rather allow for a number of different implementations.

(I'm not saying that you _have_ to have a class, just that it's worth keeping in mind that even though the example solution only has one method and doesn't carry state, it's possible to solve this in a way that would have multiple methods and carry state.)

Reply:
I agree, as indicated with my approval for this PR (I could've been more explicit). There is nothing wrong with this example implementation, there is no right way to do it. My focus of the review was primarily the _test cases as a means to communicate the problem spec to the learner while being implementation agnostic_. 
However since the example implementation needs to pass the test cases (in CI) to prove the correctness of the test cases themselves, I am inclined towards a solution that passes the tests but also makes the test cases not assume much about the implementation. 

This is the interesting problem to solve here, because dart has a sound type system, because we have an example implementation, because we run the tests in CI against that example. How would we write the tests in a way that is sound for the type system while assuming the least about the implementation.

#### Considering that, I propose another alternative, implement as a top-level function named `toPigLatin`
```dart
// test/pig_latin_test.dart

import 'package:pig_latin/pig_latin.dart';

void main() {
  // clarity on usage sites is preserved.
  expect(toPigLatin('apple'), equals('appleay'));
  
  // Notice that I don't save the result of that function in a variable first and then compare it. I do that inline.
  // The assertion statement is crafted this way to remove a lot of assumptions regarding the implementation. See below
}

```

#### The implementation can take various many forms,
```dart
// lib/pig_latin.dart

class PigLatin1 {
  String translate(String phrase) {}
}

/// First form, if they want to use a class based implementation
final toPigLatin = PigLatin1().translate;

class PigLatin2 {
   /// This makes the object of this class act as a function
   String call(String phrase) {}
}

/// 2nd Form, a fancy class pretending to be a function
final toPigLatin = PigLatin2(); // const is also possible here


/// 3rd Form, top-level function
String toPigLatin(String phrase) {}


/// 4th Form
/// A top-level getter that returns a function
String Function(String phrase) get toPigLatin => {
  return (phrase) => PigLatin1().translate(phrase);
}


/// 5th Form, for those that live dangerously 😉 
class toPigLatin {
  String _translatedPhrase;

  /// This is the constructor, not a function
  toPigLatin(String phrase) {
   _translatedPhrase = // TODO: actually translate(phrase);
  }

  operator ==(dynamic expected) {
    return expected is String && _translatedPhrase == expected;
  }
  
  String toString() => _translatedPhrase;
}

/// The above when printed would act as a string
print(toPigLatin('apple')) // 'appleay'
// It can even be compared with strings (equality)
print(toPigLatin('apple') == 'appleay'); // true

// And because the testing library by default uses the object's `==` operator for equality checks
// Following test assertion would also work
expect('appleay', equals(toPigLatin('apple')));  // works as expected, ref: https://github.com/dart-lang/matcher/blob/6a9b83bbd73e50df2058b3e8e4aa301df49569c6/lib/src/equals_matcher.dart#L183

// However we don't have to rely on the implementation of the test matcher, we can also 
expect(toPigLatin('apple').toString(), equals('appleay')); // This form is essentially allowing any return type from the toPigLatin invocation, as long as it's stringy value equals 'appleay'

// Since `toString()` is present on all objects, it's type safe to call it on anything.
// Since when you call `toString()` on an actual String, it's  NOOP.

// You could also write the equals function at the top of the test file as,
bool Function(Object?) equals(String expected) => (actual) => actual == expected; // ref: https://github.com/dart-lang/matcher/blob/6a9b83bbd73e50df2058b3e8e4aa301df49569c6/lib/src/util.dart#L40-L42

// And use it as you would normally
expect(toPigLatin('apple'), equals('appleay'));


/// There's a couple more I can write, but I think the point is clear.

```

In all of these forms, usage sites will remain the same, i.e., You'd use them exactly the same way
`toPigLatin('apple'))`

---

The test cases only care about a symbol `toPigLatin` being made publicly available from the implementation and it has the type definition `Function(String phrase)` _upon it's invocation_. Notice that there is no return type in that type definition. Since we really care about it being _equal_ to the String we match it with, which can be accomplished in many ways. The symbol `toPigLatin` does not need to be a function all the time, just _some times_ it needs to act like one. And it doesn't need to return a string, the returned value needs to be _equal_ to a string, only sometimes.

I know I made the 5th form a little bit more complicated than it needs to be. But it removes a lot of assumptions about the implementation. 
- Does `toPigLatin` needs to be a function? No!
- Does it need to be a class? No!
- Does it need to be an object? No!
- Can it be a mix of all or none of all? Yes!.
- Does it absolutely need to return a string? No!
- Does it need to return anything at all? No, but actually Yes!
- Can `toPigLatin` be just an `Object`? Yes! Like a literal `Object`? Yes!
  ```dart
  const toPigLatin = Object();

  extension PigLatinTranslator on Object {
    String call(String phrase) {
      // TODO: implement translation logic
    }
  }

  /// The above makes it possible to do this, even though toPigLatin is literally an object.
  print(toPigLatin('apple') == 'appleay') // true
  ```
- Can `toPigLatin` be `null` and still pass all the tests? I won't tell you the answer to that! Try it yourself! 😃 



---

Once you write it as a class `PigLatin` with a `translate` function, import, instantiate an object & use it in tests, you're defining it's public interface. The first thought that comes to mind when I need to solve a problem that has the following code, is _Oh, I need a class named PigLatin with a translate function_. But that's just me.
```dart
final piglatin = PigLatin();
expect(piglatin.translate('apple'), equals('appleay'));
```

The same can be said for the following, my first thought would be _Oh, I need a function named toPigLatin_. But hey!, that's just me. :)
```dart
expect(toPigLatin('apple'), equals('appleay'));
```

---

I'd like to re-emphasize that, this is just for the sake of the conversation, your implementation is perfectly fine. Regardless of which approach you chose, it will inevitably influence how the tests are written, which will influence how it's going to be implemented. However, if you reverse the process, i.e. you write the tests first and then chose an implementation (the whole point of TDD). It helps you remove a lot of assumptions.

Reply:
Thanks so much, that was really helpful. I don't know anything about Dart (yet), so I'm only commenting based on what I infer from the conversation.